src/Microsoft.Health.Dicom.Core/Extensions/DicomDatasetExtensions.cs

```diff
@@ -158,5 +159,37 @@ public static void AddValueIfNotNull<T>(this DicomDataset dicomDataset, DicomTag
                 dicomDataset.Add(dicomTag, value);
             }
         }
+
+        /// <summary>
+        /// Get custom tag indexes
+        /// </summary>
+        /// <param name="dicomDataset">The dicom dataset.</param>
+        /// <param name="customTagStoreEntries">The custom tag entries stored.</param>
+        /// <returns>The indexes.</returns>
+        public static Dictionary<long, DicomItem> GetCustomTagIndexes(this DicomDataset dicomDataset, IEnumerable<CustomTagStoreEntry> customTagStoreEntries)
+        {
+            // We only support index on top level custom tag for now, will support embeded tag later.
+            Dictionary<long, DicomItem> indexes = new Dictionary<long, DicomItem>();
+            Dictionary<string, CustomTagStoreEntry> tagPathDictionary = customTagStoreEntries.ToDictionary(keySelector: item => item.Path, comparer: StringComparer.OrdinalIgnoreCase);
```

Comment:
Would it make sense to do this conversion a few layers up and just pass the dictionary here? My reading of the code seems like we'll be doing this work many times and will always get the same results.

Reply:
In looking further up the stack, it appears that we'll be doing this for every single instance.